modules/kvformat/filterx-func-parse-kv.c

```diff
@@ -0,0 +1,384 @@
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx-func-parse-kv.h"
+#include "filterx/object-string.h"
+#include "filterx/object-primitive.h"
+#include "filterx/expr-literal.h"
+#include "filterx/filterx-eval.h"
+#include "filterx/filterx-globals.h"
+#include "filterx/object-json.h"
+#include "filterx/object-message-value.h"
+#include "filterx/object-null.h"
+
+#include "kv-parser.h"
+#include "parser/parser-expr.h"
+#include "scratch-buffers.h"
+
+#define PARSE_KV_OPTS_COUNT 4
+
+enum parse_kv_opts
+{
+  PARSE_KV_OPTS_PREFIX = 0,
+  PARSE_KV_OPTS_VALUE_SEP = 1,
+  PARSE_KV_OPTS_PAIR_SEP = 2,
+  PARSE_KV_OPTS_STRAY_WORDS_VAL_NAME = 3,
+};
+
+static const gchar *parse_kv_opts_names[PARSE_KV_OPTS_COUNT] =
+{
+  "prefix",
+  "value_separator",
+  "pair_separator",
+  "stray_words_value_name",
+};
+
+typedef struct FilterXFunctionParseKV_
+{
+  FilterXFunction super;
+  FilterXExpr *msg;
+  gchar value_separator;
+  gchar *pair_separator;
+  gchar *prefix;
+  gchar *stray_words_value_name;
+  gsize prefix_len;
+  KVScanner scanner;
+} FilterXFunctionParseKV;
+
+static gboolean
+_is_valid_separator_character(char c)
+{
+  return (c != ' '  &&
+          c != '\'' &&
+          c != '\"' );
+}
+
+static void
+_set_prefix(FilterXFunctionParseKV *self, const gchar *prefix)
+{
+  // KVParser *self = (KVParser *)p;
+
+  g_free(self->prefix);
+  if (prefix)
+    {
+      self->prefix = g_strdup(prefix);
+      self->prefix_len = strlen(prefix);
+    }
+  else
+    {
+      self->prefix = NULL;
+      self->prefix_len = 0;
+    }
+}
+
+static void
+_set_value_separator(FilterXFunctionParseKV *self, gchar value_separator)
+{
+  self->value_separator = value_separator;
+}
+
+static void
+_set_pair_separator(FilterXFunctionParseKV *self, const gchar *pair_separator)
+{
+  g_free(self->pair_separator);
+  self->pair_separator = g_strdup(pair_separator);
+}
+
+static void
+_set_stray_words_value_name(FilterXFunctionParseKV *self, const gchar *value_name)
+{
+  g_free(self->stray_words_value_name);
+  self->stray_words_value_name = g_strdup(value_name);
+}
+
+static const gchar *
+_get_formatted_key_with_prefix(FilterXFunctionParseKV *self, const gchar *key, GString *formatted_key)
+{
+  if (formatted_key->len > 0)
+    g_string_truncate(formatted_key, self->prefix_len);
+  else
+    g_string_assign(formatted_key, self->prefix);
+  g_string_append(formatted_key, key);
+  return formatted_key->str;
+}
+
+static const gchar *
+_get_formatted_key(FilterXFunctionParseKV *self, const gchar *key, GString *formatted_key)
+{
+  if (!self->prefix)
+    return key;
+  return _get_formatted_key_with_prefix(self, key, formatted_key);
+}
+
+static gboolean
+_apply_parse_kv_option(FilterXFunctionParseKV *self, int opt, const gchar *opt_val, GError **error)
```

Comment:
[optional]

To me it is a bit strange to see that `_extract_parse_kv_opts()` returns a ptr array, which is only used here. I feel like the eval + type check and setting fields of self could be done in one function.

Reply:
This was resolved in the meanwhile